docs: add note on common misconceptions about type narrowing (#7391)
typescript-eslint/typescript-eslint: packages/eslint-plugin/docs/rules/no-unnecessary-condition.md

].filter(t => t); // number[] is always truthy
```

:::note

> Currently in TypeScript, function calls have no influence on type narrowing.

Reviewer: Add a link to microsoft/TypeScript#9998

Author: Done

Even if a function has side effects that modify the condition, it isn't taken into account.
As a result, code like this violates the rule.

```ts
let condition = false;

const modifyCondition = () => {
  condition = true;
};

modifyCondition();

> [(1, 2, 3)].forEach(modifyCondition);

Reviewer: ```suggestion
[1, 2, 3].forEach(modifyCondition);
```

].filter(t => t); // number[] is always truthy
```

> :::note

Reviewer: I suggest moving it out of "examples", because the content is turned into tabs so we want each section to be approximately equal height. My immediate thought is to add it as a bullet under "When not to use it" and explain that if you run into `#9998` you'll likely want to `eslint-disable-next-line`.

The main downside to using this rule is the need for type information.

This rule has a known edge case of triggering on conditions that were modified within function calls (as side effects).
It is due to limitations of TypeScript's type narrowing.

> We recommend you use an [inline ESLint disable comment](https://eslint.org/docs/latest/use/configure/rules#using-configuration-comments-1).

Reviewer: Actually, to be exact you don't need disable to work around `#9998`. For example:

```js
let x = false;

const f = () => (x = true);
f();

if (x) console.log();
```

You just need an upcast `false as boolean`.

The main downside to using this rule is the need for type information.

This rule has a known edge case of triggering on conditions that were modified within function calls (as side effects).
It is due to limitations of TypeScript's type narrowing.
See [#9998](https://github.com/microsoft/TypeScript/issues/9998) for details.

> We recommend to upcast the condition to `boolean` with a [Type Assertion](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions).
> 
> ```ts
> let condition = false;
> 
> const f = () => (condition = true);
> f();
> 
> if (condition as boolean) {
> }

Reviewer: ```suggestion
We recommend to upcast the variable with a [type assertion](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions).

```ts
let condition = false as boolean;

const f = () => (condition = true);
f();

if (condition) {
}
```

Note that `let condition: boolean = false;` will not work, as TS will still overzealously narrow `condition` to `false` based on control flow analysis.
